minio/signer.py

```diff
@@ -92,8 +92,15 @@ def presign_v4(method, url, credentials,
     if request_date is None:
         request_date = datetime.utcnow()
 
-    parsed_url = urlsplit(url)
+    # If a sha256sum is known, add to headers to include with signature
     content_hash_hex = _UNSIGNED_PAYLOAD
+    for k in headers.keys():
+        if k.lower() == 'x-amz-content-sha256':
+            content_hash_hex = headers.get(k, _UNSIGNED_PAYLOAD)
```

Comment:
I'm rebasing some changes on-top of this and wondered about this snippet.

Is there is a reason to call `keys()` (given iterating on a dict yields the keys anyway), and secondly the `headers.get` method call has a default of `_UNSIGNED_PAYLOAD` which appears to be unreachable by definition?

Reply:
keys() (my two cents) is mostly helpful for the developer reader the code to see that we are iterating through a dictionary. And I agree given that we find k as a key, the second value is unreachable. I think @harshavardhana wrote this snippet so maybe he can comment. It might just be another message for the developer, in code, to hint at what the value should be?

Reply:
why would it be unreachable? @hardbyte the reason it is done to ensure that we are case insensitive, otherwise we may look at the key which may be defined with a different canonical representation in the input headers dictionary. At the same time, we don't change what user-provided as input. 

We can definitely remove the `header.keys()`

Reply:
I think what we might want instead is:
```python
content_hash_hex = headers.get(k) or _UNSIGNED_PAYLOAD
```
This will mean that if the key is set but the value is None, we default to the unsigned payload. Currently, if the header is set (and it's None or empty string) we then use that empty value. What would be the desired outcome?

Reply:
Another edge error would be having that header set, but with different capitalization. Here is a dummy example:

```
 for k in mydict.keys(): 
    if k.lower() == 'a': 
        print(mydict[k]) 

a
A
```
In this case we would get whichever is last in the matching set.

Reply:
Thanks for revisiting this.

@harshavardhana the `_UNSIGNED_PAYLOAD` is unreachable as far as I can tell - as `k` is definitely in `headers` - the default value won't ever apply.

In the first snippet @vsoch posted the default value does get used if the user has passed in something falsey (e.g. empty string or None). Perhaps passing in a falsey header should just raise a `ValueError` instead though.

Reply:
> Thanks for revisiting this.
> 
> @harshavardhana the `_UNSIGNED_PAYLOAD` is unreachable as far as I can tell - as `k` is definitely in `headers` - the default value won't ever apply.
> 
> In the first snippet @vsoch posted the default value does get used if the user has passed in something falsey (e.g. empty string or None). Perhaps passing in a falsey header should just raise a `ValueError` instead though.

Feel free to change it to be more pythonic, with consideration of case sensitivity.